File: solutions/problem_1383_maximum_performance_of_a_team/solution.cpp
/*
 * Problem 1383: Maximum Performance of a Team
 * =========================================
 * Difficulty: Hard
 * Tags: Array, Greedy, Sorting, Heap (Priority Queue)
 * Pattern: Heap / Priority Queue
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <queue>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int maxPerformance(int n, vector<int>& speed, vector<int>& efficiency, int k) {
        // Heap/Priority Queue - O(n log k) time
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int val : n) {
            pq.push(val);
            if ((int)pq.size() > speed)
                pq.pop();
        }
        return pq.empty() ? 0 : pq.top();
    }
};

File: solutions/problem_1383_maximum_performance_of_a_team/solution.py
"""
Problem 1383: Maximum Performance of a Team
=========================================
Difficulty: Hard
Tags: Array, Greedy, Sorting, Heap (Priority Queue)
Pattern: Heap / Priority Queue

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def maxPerformance(self, n: int, speed: List[int], efficiency: List[int], k: int) -> int:
        # Heap/Priority Queue - O(n log k) time
        import heapq
        if not n:
            return 0
        # Min heap (negate for max heap)
        heap = []
        for val in n:
            heapq.heappush(heap, val)
            if len(heap) > (speed if isinstance(speed, int) else len(n)):
                heapq.heappop(heap)
        return heap[0] if heap else 0
